src/internal/task/etcd_queue.go

```diff
@@ -0,0 +1,113 @@
+package task
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+
+	"github.com/cevaris/ordered_map"
+)
+
+const (
+	waitTime = 10 * time.Millisecond
+	bufSize  = 10
+)
+
+type taskFunc func()
+
+type groupEntry struct {
+	cancel       context.CancelFunc
+	taskFuncChan chan taskFunc
+}
+
+type taskQueue struct {
+	groups *ordered_map.OrderedMap
+	mu     sync.Mutex
+}
+
+func newTaskQueue(ctx context.Context) *taskQueue {
+	tq := &taskQueue{
+		groups: ordered_map.NewOrderedMap(),
+	}
+	go func() {
+		var cycleKey string
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			// Get the next group.
+			tq.mu.Lock()
+			iter := tq.groups.IterFunc()
+			kv, ok := iter()
+			tq.mu.Unlock()
+			var key string
+			if ok {
+				key = kv.Key.(string)
+			}
+			// Wait if there are no group entries or a cycle has been completed without finding a task.
+			if !ok || key == cycleKey {
+				time.Sleep(waitTime)
```

Comment:
This is probably fine since the wait time is so short, but generally this should be a select from a ticker and `ctx.Done()`